File: Count the Number of Arrays with K Matching Adjacent Elements/Count the Number of Arrays with K Matching Adjacent Elements.py
# Solution for Count the Number of Arrays with K Matching Adjacent Elements in PY

class Solution:
    MOD = 10**9 + 7

    def modPow(self, base: int, exp: int) -> int:
        result = 1
        base %= self.MOD
        while exp > 0:
            if exp % 2 == 1:
                result = (result * base) % self.MOD
            base = (base * base) % self.MOD
            exp //= 2
        return result

    def countGoodArrays(self, n: int, m: int, k: int) -> int:
        fact = [1] * (n + 1)
        invFact = [1] * (n + 1)
        for i in range(2, n + 1):
            fact[i] = (fact[i - 1] * i) % self.MOD
        for i in range(n + 1):
            invFact[i] = self.modPow(fact[i], self.MOD - 2)

        equalPairsWays = (fact[n - 1] * invFact[k] % self.MOD) * invFact[n - k - 1] % self.MOD
        assignWays = m * self.modPow(m - 1, n - k - 1) % self.MOD
        return equalPairsWays * assignWays % self.MOD

File: Count the Number of Arrays with K Matching Adjacent Elements/Count the Number of Arrays with K Matching Adjacent Elements.java
// Solution for Count the Number of Arrays with K Matching Adjacent Elements in JAVA

class Solution {
    final int MOD = 1_000_000_007;

    int modPow(long base, long exp) {
        long result = 1;
        base %= MOD;
        while (exp > 0) {
            if ((exp & 1) == 1)
                result = (result * base) % MOD;
            base = (base * base) % MOD;
            exp >>= 1;
        }
        return (int) result;
    }

    public int countGoodArrays(int n, int m, int k) {
        int[] fact = new int[n + 1];
        int[] invFact = new int[n + 1];
        fact[0] = invFact[0] = 1;
        for (int i = 1; i <= n; i++) {
            fact[i] = (int)((1L * fact[i - 1] * i) % MOD);
        }
        for (int i = 0; i <= n; i++) {
            invFact[i] = modPow(fact[i], MOD - 2);
        }

        int equalPairsWays = (int)((1L * fact[n - 1] * invFact[k] % MOD * invFact[n - k - 1]) % MOD);
        long assignWays = (1L * m * modPow(m - 1, n - k - 1)) % MOD;
        return (int)((1L * equalPairsWays * assignWays) % MOD);
    }
}
